### ISSUE TITLE
Make socket alive check optional when initializing runtime

### ISSUE THREAD

**Issue by shadyabhi** [haproxytech/client-native#97]
runtime initialization: [here](https://github.com/haproxytech/client-native/blob/master/runtime/interface.go#L150)

which ultimately probes socket to check if its alive here: [here](https://github.com/haproxytech/client-native/blob/master/runtime/runtime_single_client.go#L67)

However, is that really essential? Ultimately, every operation on socket is done via this [function](https://github.com/haproxytech/client-native/blob/master/runtime/runtime_single_client.go#L92C34-L92C34) which is a full sequence of `Dial/Write/Close`. That is great, because this way, if HAproxy restarts or crashes while the client-native runtime object is still alive in another process, it can gracefully recover from it as we do a full `Dial/Write/Close` unless the socket path changed etc, that's not in scope for this discussion.

Now, with that thought in mind, why is checking `aliveness` a hard requirement under `runtime.Init`? At best, we should make it optional with default as `check enabled`. 

Thoughts?

**Comment by oktalz** [comment]
hi @shadyabhi,

yes, there is a reason, we have some runtime operations that depends on a HAProxy version, and are implemented differently due to the version, so before starting and executing those commands, check is made and version is fetched so client can respond to that (and not do that check it every time)

having said that, we could add additional option where you could specify HAProxy version, then we could remove this check.
this is something that could be done.

**Comment by shadyabhi** [comment]
@oktalz Thank you for the reply. 

>yes, there is a reason, we have some runtime operations that depends on a HAProxy version, and are implemented differently due to the version, so before starting and executing those commands, check is made and version is fetched so client can respond to that (and not do that check it every time)

And if the socket is unavailable, the version check should fail as well? So, what's the point of sending `help` command at runtime initialization, at best, it'll be great if it can be made optional. 

>having said that, we could add additional option where you could specify HAProxy version, then we could remove this check. this is something that could be done.

Do you have any suggestions on how you want to approach this? Would you approach it by adding a new field in `option.RuntimeOptions`? 
Since this is a `socket specific` info, it could also be possible to put a new field in `option.Sockets`, but that'll break the API. 

I'll be curious to hear your thoughts.

**Comment by shadyabhi** [comment]
I should add more context. At my company, we've a separate process which plays the role of `controlplane`for HAproxy deployment. This `controlplane` has a completely independent deployment lifecycle. We want to make sure that we should be able to deploy `controlplane` before `haproxy` is ready.  Today, we're are able to achieve this by initializing `clientnative.HAproxyClient` without runtime and then using the `ReplaceRuntime method call` to configure runtime when HAproxy is ready.

Since, `client-native` is designed in a way that each `executeRaw` call does a fresh `Dial/Write/Close`, I felt that it is reasonable to make the `help check` on `runtime initialization` optional.

**Comment by oktalz** [comment]
hi @shadyabhi 
>In what manner, I can answer in few days.

sorry for that, it took a little bit more time :(

now we have two options:
https://github.com/haproxytech/client-native/blob/master/runtime/options/do-not-check-runtime-on-init.go
https://github.com/haproxytech/client-native/blob/master/runtime/options/allow-delayed-start.go

with first one you can completely remove the check, with second one some time is allowed before returning an error

beside master, this was also backported to `v5.1` branch

**Comment by shadyabhi** [comment]
@oktalz This is great, thanks for sharing this, exactly what we needed. There is a valid use-case for this when the control-plane is independently deployed from haproxy. 

Since, everything's addressed, I'll go ahead and close this.
